[Bug] hs_read_persisted_state unconditionally restores via nameref, creating globals and overwriting pre-set locals

## Summary

The explicit restore path in `hs_read_persisted_state` (lines 458-461 of `config/handle_state.sh`) assigns through a nameref unconditionally, without checking whether the target variable is actually declared local in the immediate caller's scope.

## Affected code

```bash
for __requested_var in "${!__restored_map[@]}"; do
    local -n __requested_var_ref="$__requested_var"
    __requested_var_ref="${__restored_map[$__requested_var]}"
done
```

## Problem

Because there is no `local -p` guard before the assignment, two violations of the design intent occur:

1. **Creates globals**: if the requested variable name is not declared as a local anywhere in the caller's scope, `local -n __requested_var_ref="$__requested_var"` causes the assignment to land in global scope.

2. **Overwrites pre-set locals**: if the requested variable is declared local in the caller but already has a non-empty value (set before the restore call), the restore unconditionally overwrites it.

The old internal snippet format had an explicit `if local -p var ...; then ... fi` guard that prevented both violations. The current nameref restore path never replicated that guard.

## Expected behaviour

The function should only restore a value into a variable that:
- is declared local in the immediate caller's scope (`local -p` succeeds), **and**
- is currently unset (to avoid silently overwriting a value the caller has already computed).

Restoring into an undeclared name should be an error (or at minimum a warning with `-q`), not a silent global creation.

## Root cause

The guard was present in the deprecated eval-snippet code path but was omitted when the nameref-based restore was written.

## Test coverage gap

All existing tests for `hs_read_persisted_state` pre-declare the target locals before calling the function, so the guard-less path was never exercised. Tests at lines 404 ("eval without local") and 429 ("eval restores onto new locals") exercise the boundary conditions but currently pass because they rely on the old `eval "$state"` internal format. When those tests are converted to the modern API they will expose this regression.

## Impact

- Any caller that passes a variable name to `hs_read_persisted_state` without pre-declaring it local will silently pollute global scope.
- Library code following the documented pattern (`local temp_file; hs_read_persisted_state "$@" -- temp_file`) is safe, but the function itself provides no enforcement.

## Proposed fix

Before the nameref assignment loop, check `local -p "$__requested_var"` in the immediate caller scope. If the check fails, emit a warning (or error under strict mode) and skip the assignment rather than creating a global.

## Issue Assessment — Task 2

### Historical context

The `return 1` in the guard snippet was not a bug that got overlooked: it was a deliberate design artifact from the original architecture.

In the original implementation `hs_read_persisted_state` was a trivial passthrough:

```bash
hs_read_persisted_state() {
    local state_string="$1"
    printf '%s' "$state_string"
}
```

The guard logic lived in `hs_persist_state_as_code`, which **generated** a Bash snippet that the caller was required to `eval`. The snippet embedded:

```bash
if local -p VAR >/dev/null 2>&1; then
  if [ -n "${VAR+x}" ] && [ -n "$VAR" ]; then
    printf "[ERROR] local VAR already defined; refusing to overwrite\n" >&2
    return 1
  else
    VAR=<value>
  fi
fi
```

Because the snippet ran inside the caller via `eval`, the `return 1` aborted the **caller** (not a nested function), which was the intended effect. There was no way to translate it into a named `HS_ERR_*` constant because the snippet had no access to those constants in the general case. The silent-skip for undeclared locals was also intentional in that design: if `local -p VAR` failed, the assignment block was simply not executed.

The nameref-based restore path introduced in commit `7a709b1` eliminated the `eval` requirement and therefore needed to reimplement the guards natively — but they were never added.

### Scope of fix

The fix must add two guards inside the `for __requested_var in "${!__restored_map[@]}"` loop (lines 457–460 of `config/handle_state.sh`):

1. **Not declared local in caller**: check `local -p "$__requested_var"` from within `hs_read_persisted_state` — which, being one frame above the caller in the nameref path, can introspect the caller's locals via `caller`-relative `local -p`. On failure, emit an error message and return `HS_ERR_UNKNOWN_VAR_NAME=10` (reuse; same semantic as the code added in #99 for `hs_persist_state_as_code`).

2. **Already set in caller**: after the `local -p` check passes, check whether the variable is already set (non-empty). On failure, emit an error message and return a new constant `HS_ERR_VAR_ALREADY_SET=11`.

### Summary of changes

| Area | Change |
|---|---|
| `config/handle_state.sh` | Add `HS_ERR_VAR_ALREADY_SET=11` constant; add two guards in the nameref restore loop |
| `docs/libraries/handle_state.rst` | Document the two new error conditions under `hs_read_persisted_state`; add `HS_ERR_VAR_ALREADY_SET=11` to the Error Codes table; fix all examples to use `\|\| return $?` |
| `test/test-hs_persist_state.bats` | Add tests for both error paths (undeclared target, pre-set target); verify correct error codes and stderr messages |

### Blockers

None. All changes are local to the three files above.

Awaiting approval to proceed to Task 3 (Implementation Planning).

## Implementation Plan — Task 3

Branch: `bug/issue-100-guard-nameref-restore`

### Design note: guard implementation strategy

Bash provides no API to introspect the local-variable frame of a specific caller from a callee. `local -p VAR` only sees the current function's locals; `declare -p VAR` traverses the full dynamic scope (caller chain + globals). This is the same constraint that forced the old implementation to embed the `if local -p …` guard inside a generated snippet that was `eval`'d in the caller's scope.

The nameref-based restore path cannot replicate the exact guard. The pragmatic equivalent, consistent with how `hs_persist_state_as_code` handles the analogous check (line 159), is:

- **Guard 1 (not declared)**: `declare -p "$VAR"` — fails if the variable is not accessible anywhere in the dynamic scope (catches the most common mistake: forgetting `local VAR` in the caller). Returns `HS_ERR_UNKNOWN_VAR_NAME=10`.
- **Guard 2 (already set)**: `${!VAR+x}` — non-empty if the variable already holds a value (local or global). Returns the new `HS_ERR_VAR_ALREADY_SET=11`.

The residual edge case — a variable that is an unset global (not local) — is not detectable without running code in the caller's scope. This is acceptable given the Bash limitation; the common mistake is fully covered.

### Files and changes

**`config/handle_state.sh`**

1. Add `readonly HS_ERR_VAR_ALREADY_SET=11` after line 48.
2. Update the `hs_read_persisted_state` header error-list to document the two new conditions.
3. In the nameref restore loop (lines 457–460), insert the two guards before the `local -n` assignment:
   ```bash
   if ! declare -p "$__requested_var" >/dev/null 2>&1; then
       echo "[ERROR] hs_read_persisted_state: '$__requested_var' is not declared in scope." >&2
       return "$HS_ERR_UNKNOWN_VAR_NAME"
   fi
   if [[ "${!__requested_var+x}" ]]; then
       echo "[ERROR] hs_read_persisted_state: '$__requested_var' is already set; refusing to overwrite." >&2
       return "$HS_ERR_VAR_ALREADY_SET"
   fi
   ```
4. Add `HS_ERR_VAR_ALREADY_SET` to the `export` list in the `hs_destroy_state` subprocess (lines 284–287).

**`docs/libraries/handle_state.rst`**

1. Add `HS_ERR_VAR_ALREADY_SET=11` to the Error Codes table.
2. Add two new bullets to the `hs_read_persisted_state` Errors section.
3. Add `|| return $?` to every `hs_read_persisted_state`, `hs_persist_state_as_code`, and `hs_destroy_state` call in all code examples.

**`test/test-hs_persist_state.bats`**

1. Add test: explicit restore into a variable not declared in caller scope → exit code equals `$HS_ERR_UNKNOWN_VAR_NAME`, stderr contains `"is not declared in scope"`.
2. Add test: explicit restore into a pre-set local → exit code equals `$HS_ERR_VAR_ALREADY_SET`, stderr contains `"is already set"`.

### Testing strategy

- Preliminary tests (Task 4): one test per new error path, written before source code changes.
- Edge-case expansion (Task 5): test that a correctly declared, unset local restores normally; test that `-q` does not suppress the errors (errors are not warnings); test with multiple requested vars that the first bad var stops processing.

Awaiting approval to proceed to Task 4 (Documentation and Preliminary Tests).

## Revised Implementation Plan — Task 3 (updated)

Branch: `bug/issue-100-guard-nameref-restore`

### Clarified design intent for the two restore forms

The two forms serve different, complementary purposes:

**Implicit form** (`eval "$(hs_read_persisted_state "$@")"`) — **preferred for the common case**
- The generated snippet runs `local -p` directly in the caller's scope, so it can only target variables that are declared local AND unset in the immediate caller.
- Provably free of global pollution: variables are discovered automatically from the caller's local frame.
- Correct choice whenever the caller wants to restore all relevant state into its own locals.

**Explicit form** (`hs_read_persisted_state "$@" -- x y z`) — **appropriate when targeting variables outside the immediate caller's local frame**
- Targets variables by name, traversing the full dynamic scope (`declare -p`).
- Intentional use cases: restoring into a variable declared in a higher-up caller, or into an explicitly declared (but unset) global.
- The `declare -p` guard is a feature here, not a limitation: it validates that the name is accessible somewhere in the dynamic scope, which is exactly what the caller intends.
- The residual edge case (unset global with a coincidentally matching name) is treated as a legitimate target, not an error, because the caller named it explicitly.

### Guard behaviour (unchanged from prior plan)

Guard 1 — not declared anywhere in dynamic scope: `declare -p "$VAR"` fails → `HS_ERR_UNKNOWN_VAR_NAME=10`.
Guard 2 — already set: `${!VAR+x}` non-empty → `HS_ERR_VAR_ALREADY_SET=11`.

### Files and changes

**`config/handle_state.sh`**

1. Add `readonly HS_ERR_VAR_ALREADY_SET=11` after line 48.
2. Update the `hs_read_persisted_state` header to document the two new error conditions and clarify when each restore form is appropriate.
3. Insert the two guards in the nameref restore loop (before lines 457–460):
   ```bash
   if ! declare -p "$__requested_var" >/dev/null 2>&1; then
       echo "[ERROR] hs_read_persisted_state: '$__requested_var' is not declared in scope." >&2
       return "$HS_ERR_UNKNOWN_VAR_NAME"
   fi
   if [[ "${!__requested_var+x}" ]]; then
       echo "[ERROR] hs_read_persisted_state: '$__requested_var' is already set; refusing to overwrite." >&2
       return "$HS_ERR_VAR_ALREADY_SET"
   fi
   ```
4. Add `HS_ERR_VAR_ALREADY_SET` to the `export` list in the `hs_destroy_state` subprocess (lines 284–287).

**`docs/libraries/handle_state.rst`**

1. Add `HS_ERR_VAR_ALREADY_SET=11` to the Error Codes table.
2. Add two new bullets to the `hs_read_persisted_state` Errors section.
3. Rewrite the usage guidance to establish the form hierarchy:
   - **Preferred**: implicit form (`eval "$(hs_read_persisted_state "$@")"`) for restoring into the immediate caller's own unset locals — zero risk of global pollution.
   - **Explicit form** (`-- x y z`): the right choice when intentionally targeting variables declared higher up in the call hierarchy, or declared globals; also useful when only a named subset is needed.
4. Add `|| return $?` to every `hs_read_persisted_state`, `hs_persist_state_as_code`, and `hs_destroy_state` call in all code examples.

**`test/test-hs_persist_state.bats`**

1. Test: explicit restore into an undeclared name → `HS_ERR_UNKNOWN_VAR_NAME`, expected stderr.
2. Test: explicit restore into a pre-set local → `HS_ERR_VAR_ALREADY_SET`, expected stderr.

### Testing strategy

- Preliminary tests (Task 4): one test per new error path, written before source code changes.
- Edge-case expansion (Task 5): correct unset local restores normally; `-q` does not suppress errors; first bad var in a multi-var list stops processing; explicit form correctly targets an unset variable declared in a grandparent scope.

Awaiting approval to proceed to Task 4 (Documentation and Preliminary Tests).

--- a/config/handle_state.sh
+++ b/config/handle_state.sh
@@ -364,7 +364,7 @@ hs_destroy_state() {
 #   - `HS_ERR_UNKNOWN_VAR_NAME` if a requested variable name (explicit form)
 #     is not declared anywhere in the dynamic scope.
 #   - `HS_ERR_VAR_ALREADY_SET` if a requested variable name (explicit form)
-#     already holds a value; unset it first if an overwrite is intended.
+#     is non-empty; unset or empty it first if an overwrite is intended.
 #   - Missing requested variables are warnings, one per variable, unless `-q`
 #     is supplied.
 # Usage examples:
@@ -474,7 +474,7 @@ hs_read_persisted_state() {
                 echo "[ERROR] hs_read_persisted_state: '$__requested_var' is not declared in scope." >&2
                 return "$HS_ERR_UNKNOWN_VAR_NAME"
             fi
-            if [[ "${!__requested_var+x}" ]]; then
+            if [[ -n "${!__requested_var}" ]]; then
                 echo "[ERROR] hs_read_persisted_state: '$__requested_var' is already set; refusing to overwrite." >&2
                 return "$HS_ERR_VAR_ALREADY_SET"
             fi

--- a/config/handle_state.sh
+++ b/config/handle_state.sh
@@ -46,6 +46,7 @@ readonly HS_ERR_STATE_VAR_UNINITIALIZED=7
 readonly HS_ERR_MISSING_ARGUMENT=8
 readonly HS_ERR_INVALID_ARGUMENT_TYPE=9
 readonly HS_ERR_UNKNOWN_VAR_NAME=10
+readonly HS_ERR_VAR_ALREADY_SET=11
 
 # --- hs_persist_state_as_code ----------------------------------------------------------
 # Function:
@@ -284,7 +285,7 @@ hs_destroy_state() {
                 export HS_ERR_RESERVED_VAR_NAME HS_ERR_VAR_NAME_COLLISION \
                     HS_ERR_MULTIPLE_STATE_INPUTS HS_ERR_CORRUPT_STATE \
                     HS_ERR_INVALID_VAR_NAME HS_ERR_STATE_VAR_UNINITIALIZED \
-                    HS_ERR_INVALID_ARGUMENT_TYPE
+                    HS_ERR_INVALID_ARGUMENT_TYPE HS_ERR_VAR_ALREADY_SET
                 declare -fx _hs_is_array _hs_is_valid_variable_name \
                     _hs_resolve_state_inputs hs_persist_state_as_code
 
@@ -469,6 +470,14 @@ hs_read_persisted_state() {
         fi
 
         for __requested_var in "${!__restored_map[@]}"; do
+            if ! declare -p "$__requested_var" >/dev/null 2>&1; then
+                echo "[ERROR] hs_read_persisted_state: '$__requested_var' is not declared in scope." >&2
+                return "$HS_ERR_UNKNOWN_VAR_NAME"
+            fi
+            if [[ "${!__requested_var+x}" ]]; then
+                echo "[ERROR] hs_read_persisted_state: '$__requested_var' is already set; refusing to overwrite." >&2
+                return "$HS_ERR_VAR_ALREADY_SET"
+            fi
             local -n __requested_var_ref="$__requested_var"
             __requested_var_ref="${__restored_map[$__requested_var]}"
         done

--- a/config/handle_state.sh
+++ b/config/handle_state.sh
@@ -364,7 +364,7 @@ hs_destroy_state() {
 #   - `HS_ERR_UNKNOWN_VAR_NAME` if a requested variable name (explicit form)
 #     is not declared anywhere in the dynamic scope.
 #   - `HS_ERR_VAR_ALREADY_SET` if a requested variable name (explicit form)
-#     is non-empty; unset or empty it first if an overwrite is intended.
+#     is set (including empty string); unset it first if an overwrite is intended.
 #   - Missing requested variables are warnings, one per variable, unless `-q`
 #     is supplied.
 # Usage examples:
@@ -474,7 +474,7 @@ hs_read_persisted_state() {
                 echo "[ERROR] hs_read_persisted_state: '$__requested_var' is not declared in scope." >&2
                 return "$HS_ERR_UNKNOWN_VAR_NAME"
             fi
-            if [[ -n "${!__requested_var}" ]]; then
+            if [[ "${!__requested_var+x}" ]]; then
                 echo "[ERROR] hs_read_persisted_state: '$__requested_var' is already set; refusing to overwrite." >&2
                 return "$HS_ERR_VAR_ALREADY_SET"
             fi

--- a/config/handle_state.sh
+++ b/config/handle_state.sh
@@ -323,33 +323,35 @@ hs_destroy_state() {
     printf -v "$__output_state_var" '%s' "$__output"
 }
 # --- hs_read_persisted_state --------------------------------------------------------
-# Function: 
+# Function:
 #   hs_read_persisted_state [options] [--] [state_variable ...]
 # Description:
 #   Restores the values of the specified local variables from the opaque state
 #   object held in the variable named by -S.
-#   With an explicit variable list: evaluates the state in an isolated
-#   subprocess, then writes each restored value back into the matching local
-#   in the caller's scope via nameref assignment.
-#   Without an explicit variable list (and no --): emits an implicit restore
-#   snippet to stdout that the caller must eval; the snippet auto-discovers
-#   unset scalar locals in the immediate caller scope and reenters
-#   hs_read_persisted_state with those names explicitly.
+#   Preferred (implicit) form — no -- and no variable names: emits a restore
+#   snippet to stdout that the caller must eval; the snippet uses local -p in
+#   the caller's scope so it can only target unset scalar locals of the
+#   immediate caller, making it provably free of global scope pollution.
+#   Explicit form — variable names supplied after --: restores each name by
+#   traversing the full dynamic scope (caller chain and globals). Use when
+#   targeting a variable in a higher-level caller or a declared global.
 #   With -- and no variable names: returns 0 without restoring anything,
-#   disabling the auto-probe path.
+#   disabling the implicit-probe path.
 # Options:
 #   -q - suppresses the warning that is normally emitted when a requested
-#        state variable is not present in the state object.
+#        state variable is not present in the state object. Does not suppress
+#        errors.
 #   -S <state> - pass the state object by name, mandatory.
 #   Other options are ignored up to the last --, so this function is usually able
 #   to directly process its caller's argument list, future-proofing it against
 #   new hs_read_persisted_state options.
 #   -- - marks the end of options and the beginning of the list of variable names.
 # Arguments:
-#   $@ - names of local variables to restore. Without `--`, the trailing
-#        arguments that are valid Bash identifiers are treated as the variable list.
-#        Note that the value associated with the last given option will be mistaken
-#        for a variable unless that option is known or `--` is used.
+#   $@ - names of variables to restore (explicit form). Without `--`, the
+#        trailing arguments that are valid Bash identifiers are treated as the
+#        variable list. Note that the value associated with the last given
+#        option will be mistaken for a variable unless that option is known or
+#        `--` is used.
 # Errors:
 #   - `HS_ERR_MISSING_ARGUMENT` if no state variable name is supplied at all.
 #   - `HS_ERR_INVALID_VAR_NAME` if the state variable name or a requested
@@ -358,12 +360,24 @@ hs_destroy_state() {
 #     the named state variable is unset or empty.
 #   - `HS_ERR_CORRUPT_STATE` if the state object cannot be evaluated safely
 #     while restoring requested variables.
+#   - `HS_ERR_UNKNOWN_VAR_NAME` if a requested variable name (explicit form)
+#     is not declared anywhere in the dynamic scope.
+#   - `HS_ERR_VAR_ALREADY_SET` if a requested variable name (explicit form)
+#     already holds a value; unset it first if an overwrite is intended.
 #   - Missing requested variables are warnings, one per variable, unless `-q`
 #     is supplied.
 # Usage examples:
+#   # Preferred: implicit form, targets only the caller's own unset locals.
 #   cleanup() {
 #       local temp_file resource_id
-#       hs_read_persisted_state "$@" -- temp_file resource_id
+#       eval "$(hs_read_persisted_state "$@")" || return $?
+#       rm -f "$temp_file"
+#       printf 'Cleaned up resource: %s\n' "$resource_id"
+#   }
+#   # Explicit form: use when targeting a specific subset or higher-scope vars.
+#   cleanup() {
+#       local temp_file resource_id
+#       hs_read_persisted_state "$@" -- temp_file resource_id || return $?
 #       rm -f "$temp_file"
 #       printf 'Cleaned up resource: %s\n' "$resource_id"
 #   }

--- a/docs/libraries/handle_state.rst
+++ b/docs/libraries/handle_state.rst
@@ -172,9 +172,8 @@ Behavior:
 
 - Each requested name is looked up by traversing the full dynamic scope.
 - A name not declared anywhere in the dynamic scope is an error.
-- A name that is non-empty is an error; unset or empty the variable explicitly
-  before calling if an overwrite is intended. An empty-string value (``local
-  foo=""``) is treated the same as an unset local and is a valid restore target.
+- A name that is set (including an empty-string value) is an error; ``unset``
+  the variable explicitly before calling if an overwrite is intended.
 - Requested names missing from the state object are warnings, one per variable.
 - ``-q`` suppresses those warnings.
 - The current implementation restores scalar string values only.
@@ -238,8 +237,8 @@ Errors:
 - ``HS_ERR_UNKNOWN_VAR_NAME=10``: a requested variable name (explicit form) is
   not declared anywhere in the dynamic scope.
 - ``HS_ERR_VAR_ALREADY_SET=11``: a requested variable name (explicit form) is
-  non-empty; the function refuses to overwrite it. An empty-string value is
-  treated the same as an unset local and is a valid restore target.
+  set (including empty string); ``unset`` the variable first if an overwrite
+  is intended.
 
 Helper API
 ----------

--- a/docs/libraries/handle_state.rst
+++ b/docs/libraries/handle_state.rst
@@ -32,20 +32,20 @@ Quick Start
    init_function() {
        local temp_file="/tmp/some_temp_file"
        local resource_id="resource_123"
-       hs_persist_state_as_code "$@" -- temp_file resource_id
+       hs_persist_state_as_code "$@" -- temp_file resource_id || return $?
    }
 
    cleanup_function() {
        local temp_file resource_id
-       hs_read_persisted_state "$@" -- temp_file resource_id
+       eval "$(hs_read_persisted_state "$@")" || return $?
        rm -f "$temp_file"
        printf 'Cleaned up resource: %s\n' "$resource_id"
-       hs_destroy_state "$@" -- temp_file resource_id
+       hs_destroy_state "$@" -- temp_file resource_id || return $?
    }
 
    local state_var=""
-   init_function -S state_var
-   cleanup_function -S state_var
+   init_function -S state_var || return $?
+   cleanup_function -S state_var || return $?
 
 Public API
 ----------
@@ -129,25 +129,52 @@ hs_read_persisted_state
 - Usage: ``hs_read_persisted_state [forwarded args] [-q] -S <statevar> [--] [var1 var2 ...]``
 - Convenience form: ``hs_read_persisted_state state_var ...`` is normalized to
   ``-S state_var ...``. Not recommended in library code; prefer explicit ``-S``.
-- Preferred usage: ``hs_read_persisted_state "$@" -- var1 var2 ...``
+
+Restore form selection
+^^^^^^^^^^^^^^^^^^^^^^
+
+Two restore forms are available. Choose based on where the target variables live:
+
+- **Implicit form** (preferred for the common case): no ``--`` and no variable
+  names are passed. The function emits a snippet that the caller ``eval``\s.
+  Because the snippet runs ``local -p`` directly in the caller's scope, it can
+  only target variables that are declared local *and* unset in the immediate
+  caller. This form is provably free of global scope pollution.
+
+  .. code-block:: bash
+
+     cleanup_function() {
+         local temp_file resource_id
+         eval "$(hs_read_persisted_state "$@")" || return $?
+         rm -f "$temp_file"
+         printf 'Cleaned up resource: %s\n' "$resource_id"
+     }
+
+- **Explicit form**: variable names are supplied after ``--``. The function
+  restores each name by traversing the full dynamic scope (caller chain and
+  globals). Use this form when targeting a variable declared in a higher-level
+  caller, or an explicitly declared but unset global. It is also appropriate
+  when only a named subset of the state is needed.
+
+  .. code-block:: bash
+
+     cleanup_function() {
+         local temp_file resource_id
+         hs_read_persisted_state "$@" -- temp_file resource_id || return $?
+         rm -f "$temp_file"
+         printf 'Cleaned up resource: %s\n' "$resource_id"
+     }
 
 Explicit restore
 ^^^^^^^^^^^^^^^^
 
-When variable names are supplied, the function restores only those names into
-the current caller scope.
-
-.. code-block:: bash
-
-   cleanup_function() {
-       local temp_file resource_id
-       hs_read_persisted_state "$@" -- temp_file resource_id
-   }
-
 Behavior:
 
-- Restoration is by name into already-declared locals in the caller scope.
-- Requested names missing from the state are warnings, one per variable.
+- Each requested name is looked up by traversing the full dynamic scope.
+- A name not declared anywhere in the dynamic scope is an error.
+- A name that already holds a value (set, whether local or global) is an error;
+  unset the variable explicitly before calling if an overwrite is intended.
+- Requested names missing from the state object are warnings, one per variable.
 - ``-q`` suppresses those warnings.
 - The current implementation restores scalar string values only.
 
@@ -155,14 +182,15 @@ Implicit local restore
 ^^^^^^^^^^^^^^^^^^^^^^
 
 When no explicit variable names are supplied and no explicit ``--`` is present,
-``hs_read_persisted_state`` emits a small safe, locally generated implicit restore snippet. The
-caller must ``eval`` the snippet using the forwarded-arguments form:
+``hs_read_persisted_state`` emits a small safe, locally generated implicit
+restore snippet. The caller must ``eval`` the snippet using the
+forwarded-arguments form:
 
 .. code-block:: bash
 
    cleanup_function() {
        local temp_file resource_id
-       eval "$(hs_read_persisted_state "$@")"
+       eval "$(hs_read_persisted_state "$@")" || return $?
        rm -f "$temp_file"
        printf 'Cleaned up resource: %s\n' "$resource_id"
    }
@@ -206,6 +234,10 @@ Errors:
   state variable is unset or empty.
 - ``HS_ERR_CORRUPT_STATE=4``: the state cannot be evaluated safely while
   restoring explicitly requested variables.
+- ``HS_ERR_UNKNOWN_VAR_NAME=10``: a requested variable name (explicit form) is
+  not declared anywhere in the dynamic scope.
+- ``HS_ERR_VAR_ALREADY_SET=11``: a requested variable name (explicit form)
+  already holds a value; the function refuses to overwrite it.
 
 Helper API
 ----------
@@ -247,6 +279,7 @@ Error Codes
 - ``HS_ERR_MISSING_ARGUMENT=8``
 - ``HS_ERR_INVALID_ARGUMENT_TYPE=9``
 - ``HS_ERR_UNKNOWN_VAR_NAME=10``
+- ``HS_ERR_VAR_ALREADY_SET=11``
 
 Known Limitations
 -----------------
@@ -267,20 +300,20 @@ Persisting and restoring a scalar:
 
    init_function() {
        local token='a b "c" $d'
-       hs_persist_state_as_code "$@" -- token
+       hs_persist_state_as_code "$@" -- token || return $?
    }
 
    cleanup_function() {
        local token
-       hs_read_persisted_state "$@" -- token
+       hs_read_persisted_state "$@" -- token || return $?
        printf '%s\n' "$token"
    }
 
 .. code-block:: bash
 
    local state_var=""
-   init_function -S state_var
-   cleanup_function -S state_var
+   init_function -S state_var || return $?
+   cleanup_function -S state_var || return $?
 
 Representing an array manually through a scalar encoding:
 
@@ -290,26 +323,29 @@ Representing an array manually through a scalar encoding:
        local -a items=("value1" "value2" "value with spaces")
        local encoded
        encoded=$(printf '%s\0' "${items[@]}" | base64 -w0)
-       hs_persist_state_as_code "$@" -- encoded
+       hs_persist_state_as_code "$@" -- encoded || return $?
    }
 
    cleanup_function() {
        local encoded
        local -a items
-       hs_read_persisted_state "$@" -- encoded
+       hs_read_persisted_state "$@" -- encoded || return $?
        mapfile -d '' -t items < <(printf '%s' "$encoded" | base64 -d)
    }
 
 .. code-block:: bash
 
    local state_var=""
-   init_function -S state_var
-   cleanup_function -S state_var
+   init_function -S state_var || return $?
+   cleanup_function -S state_var || return $?
 
 Caveats
 -------
 
-- Prefer explicit restore lists over implicit local restore.
+- Prefer the implicit restore form (``eval "$(hs_read_persisted_state "$@")"``
+  ``|| return $?``) for cleanup functions that restore into their own locals.
+  Use the explicit form only when targeting variables in a higher-level caller
+  or declared globals.
 - Do not rely on the opaque state format being executable code forever.
 - The current implementation uses ``eval`` internally; state should therefore
   be treated as trusted input.

--- a/docs/libraries/handle_state.rst
+++ b/docs/libraries/handle_state.rst
@@ -172,8 +172,9 @@ Behavior:
 
 - Each requested name is looked up by traversing the full dynamic scope.
 - A name not declared anywhere in the dynamic scope is an error.
-- A name that already holds a value (set, whether local or global) is an error;
-  unset the variable explicitly before calling if an overwrite is intended.
+- A name that is non-empty is an error; unset or empty the variable explicitly
+  before calling if an overwrite is intended. An empty-string value (``local
+  foo=""``) is treated the same as an unset local and is a valid restore target.
 - Requested names missing from the state object are warnings, one per variable.
 - ``-q`` suppresses those warnings.
 - The current implementation restores scalar string values only.
@@ -236,8 +237,9 @@ Errors:
   restoring explicitly requested variables.
 - ``HS_ERR_UNKNOWN_VAR_NAME=10``: a requested variable name (explicit form) is
   not declared anywhere in the dynamic scope.
-- ``HS_ERR_VAR_ALREADY_SET=11``: a requested variable name (explicit form)
-  already holds a value; the function refuses to overwrite it.
+- ``HS_ERR_VAR_ALREADY_SET=11``: a requested variable name (explicit form) is
+  non-empty; the function refuses to overwrite it. An empty-string value is
+  treated the same as an unset local and is a valid restore target.
 
 Helper API
 ----------

--- a/test/test-hs_persist_state.bats
+++ b/test/test-hs_persist_state.bats
@@ -480,7 +480,7 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
     local state=""
     init state || return $?
     local foo
-    # foo is declared; bar is not — must error on bar or foo (iteration order undefined)
+    # only bar is not declared — must error on bar
     hs_read_persisted_state -S state -- foo bar || return $?
   }
   run -"$HS_ERR_UNKNOWN_VAR_NAME" --separate-stderr f
@@ -490,13 +490,11 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
 # bats test_tags=hs_read_persisted_state
 @test "hs_read_persisted_state explicit form targets unset var in grandparent scope" {
   # shellcheck disable=SC2329
+  init()  { local outer_var=from_init; hs_persist_state_as_code -S "$1" outer_var || return $?; }
+  inner() { hs_read_persisted_state -S "$1" -- outer_var || return $?; }
+  middle() { inner "$1" || return $?; }
   f() {
     local outer_var
-    middle() {
-      inner() { hs_read_persisted_state -S "$1" -- outer_var || return $?; }
-      inner "$1" || return $?
-    }
-    init() { local outer_var=from_init; hs_persist_state_as_code -S "$1" outer_var || return $?; }
     local state=""
     init state || return $?
     middle state || return $?

--- a/test/test-hs_persist_state.bats
+++ b/test/test-hs_persist_state.bats
@@ -484,7 +484,7 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
     hs_read_persisted_state -S state -- foo bar || return $?
   }
   run -"$HS_ERR_UNKNOWN_VAR_NAME" --separate-stderr f
-  [[ "$stderr" == *"is not declared in scope"* ]]
+  [[ "$stderr" == *"'bar' is not declared in scope"* ]]
 }
 
 # bats test_tags=hs_read_persisted_state

--- a/test/test-hs_persist_state.bats
+++ b/test/test-hs_persist_state.bats
@@ -400,7 +400,7 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
   [[ "$stderr" == *"'__options' conflicts with a local variable name"* ]]
 }
 
-# bats test_tags=xfail,hs_read_persisted_state
+# bats test_tags=hs_read_persisted_state
 @test "hs_read_persisted_state errors on undeclared restore target" {
   # shellcheck disable=SC2329
   f() {
@@ -414,7 +414,7 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
   [[ "$stderr" == *"is not declared in scope"* ]]
 }
 
-# bats test_tags=xfail,hs_read_persisted_state
+# bats test_tags=hs_read_persisted_state
 @test "hs_read_persisted_state errors on pre-set restore target" {
   # shellcheck disable=SC2329
   f() {
@@ -422,7 +422,7 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
     local state=""
     init state || return $?
     local baz=old
-    # baz is declared local and already set — must error, not silently overwrite
+    # baz is declared local and non-empty — must error, not silently overwrite
     hs_read_persisted_state -S state -- baz
   }
   run -"$HS_ERR_VAR_ALREADY_SET" --separate-stderr f
@@ -441,6 +441,54 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
   }
   run -0 f
   [ "$output" = "secret:v2:new" ]
+}
+
+# bats test_tags=hs_read_persisted_state
+@test "hs_read_persisted_state -q does not suppress guard errors" {
+  # shellcheck disable=SC2329
+  f() {
+    init() { local bar=v2; hs_persist_state_as_code -S "$1" bar || return $?; }
+    local state=""
+    init state || return $?
+    # -q suppresses missing-variable warnings but must not suppress guard errors
+    hs_read_persisted_state -q -S state -- bar
+  }
+  run -"$HS_ERR_UNKNOWN_VAR_NAME" --separate-stderr f
+  [[ "$stderr" == *"is not declared in scope"* ]]
+}
+
+# bats test_tags=hs_read_persisted_state
+@test "hs_read_persisted_state stops at first bad var in multi-var restore" {
+  # shellcheck disable=SC2329
+  f() {
+    init() { local foo=a bar=b; hs_persist_state_as_code -S "$1" foo bar || return $?; }
+    local state=""
+    init state || return $?
+    local foo
+    # foo is declared; bar is not — must error on bar or foo (iteration order undefined)
+    hs_read_persisted_state -S state -- foo bar || return $?
+  }
+  run -"$HS_ERR_UNKNOWN_VAR_NAME" --separate-stderr f
+  [[ "$stderr" == *"is not declared in scope"* ]]
+}
+
+# bats test_tags=hs_read_persisted_state
+@test "hs_read_persisted_state explicit form targets unset var in grandparent scope" {
+  # shellcheck disable=SC2329
+  f() {
+    local outer_var
+    middle() {
+      inner() { hs_read_persisted_state -S "$1" -- outer_var || return $?; }
+      inner "$1" || return $?
+    }
+    init() { local outer_var=from_init; hs_persist_state_as_code -S "$1" outer_var || return $?; }
+    local state=""
+    init state || return $?
+    middle state || return $?
+    printf "%s" "$outer_var"
+  }
+  run -0 --separate-stderr f
+  [ "$output" = "from_init" ]
 }
 
 # bats test_tags=hs_read_persisted_state

--- a/test/test-hs_persist_state.bats
+++ b/test/test-hs_persist_state.bats
@@ -400,39 +400,43 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
   [[ "$stderr" == *"'__options' conflicts with a local variable name"* ]]
 }
 
-# bats test_tags=hs_persist_state_as_code
-@test "eval without local should succeed and leave globals unchanged" {
+# bats test_tags=xfail,hs_read_persisted_state
+@test "hs_read_persisted_state errors on undeclared restore target" {
   # shellcheck disable=SC2329
   f() {
-    init() { local bar=v2; local baz=new; hs_persist_state_as_code -S "$1" bar baz; }
-    state=""
-    init state
-    baz=old
-    eval "$state"
-    printf "%s:%s" "$bar" "$baz"
+    init() { local bar=v2; hs_persist_state_as_code -S "$1" bar || return $?; }
+    local state=""
+    init state || return $?
+    # bar is not declared as local in f — must error, not silently create a global
+    hs_read_persisted_state -S state -- bar
   }
-  run -0 f
-  # Succeeds but does not set variables
-  [[ "$output" == ":old" ]]
-  g() {
-    init(){ local bar=v2; local baz=new; hs_persist_state_as_code -S "$1" bar baz; }
-    state=""
-    init state
-    baz=old
-    eval "$state" 2>/dev/null || true
-    [ -z "${bar+set}" ] && [ "${baz}" = "old" ]
-  }
-  run -0 g
+  run -"$HS_ERR_UNKNOWN_VAR_NAME" --separate-stderr f
+  [[ "$stderr" == *"is not declared in scope"* ]]
 }
 
-# bats test_tags=hs_persist_state_as_code
-@test "cleanup declares local and eval restores values onto new locals" {
+# bats test_tags=xfail,hs_read_persisted_state
+@test "hs_read_persisted_state errors on pre-set restore target" {
   # shellcheck disable=SC2329
   f() {
-    init(){ local foo=secret; local bar=v2; local baz=new; hs_persist_state_as_code -S "$1" foo bar baz; }
-    cleanup(){ local -n state_ref="$1"; local foo bar baz; eval "$state_ref"; printf "%s:%s:%s" "$foo" "$bar" "$baz"; }
-    state=""
-    init state
+    init() { local baz=new; hs_persist_state_as_code -S "$1" baz || return $?; }
+    local state=""
+    init state || return $?
+    local baz=old
+    # baz is declared local and already set — must error, not silently overwrite
+    hs_read_persisted_state -S state -- baz
+  }
+  run -"$HS_ERR_VAR_ALREADY_SET" --separate-stderr f
+  [[ "$stderr" == *"is already set"* ]]
+}
+
+# bats test_tags=hs_read_persisted_state
+@test "hs_read_persisted_state restores explicitly listed unset locals" {
+  # shellcheck disable=SC2329
+  f() {
+    init()    { local foo=secret bar=v2 baz=new; hs_persist_state_as_code -S "$1" foo bar baz || return $?; }
+    cleanup() { local foo bar baz; hs_read_persisted_state -S "$1" -- foo bar baz || return $?; printf "%s:%s:%s" "$foo" "$bar" "$baz"; }
+    local state=""
+    init state || return $?
     cleanup state
   }
   run -0 f

--- a/test/test-hs_persist_state.bats
+++ b/test/test-hs_persist_state.bats
@@ -422,7 +422,22 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
     local state=""
     init state || return $?
     local baz=old
-    # baz is declared local and non-empty — must error, not silently overwrite
+    # baz is declared local and set — must error, not silently overwrite
+    hs_read_persisted_state -S state -- baz
+  }
+  run -"$HS_ERR_VAR_ALREADY_SET" --separate-stderr f
+  [[ "$stderr" == *"is already set"* ]]
+}
+
+# bats test_tags=hs_read_persisted_state
+@test "hs_read_persisted_state errors on empty-string restore target" {
+  # shellcheck disable=SC2329
+  f() {
+    init() { local baz=new; hs_persist_state_as_code -S "$1" baz || return $?; }
+    local state=""
+    init state || return $?
+    local baz=""
+    # baz is set to empty string — still set, must error
     hs_read_persisted_state -S state -- baz
   }
   run -"$HS_ERR_VAR_ALREADY_SET" --separate-stderr f
@@ -529,7 +544,7 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
     init(){ local foo=secret; local bar=v2; local baz=new; hs_persist_state_as_code -S "$1" foo bar baz; }
     cleanup(){
       local state_var="$1"
-      local foo="" bar="" baz=""
+      local foo bar baz
       hs_read_persisted_state -S "$state_var" foo baz
       printf "%s:%s:%s" "$foo" "${bar:-}" "$baz"
     }
@@ -562,10 +577,10 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
   f() {
     init(){ local foo=secret; hs_persist_state_as_code -S "$1" foo; }
     outer(){
-      local foo=""
+      local foo
       inner_auto
       printf "%s:" "$foo"
-      foo=""
+      unset foo
       inner_explicit
       printf "%s" "$foo"
     }
@@ -591,7 +606,7 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
     init(){ local foo=secret; hs_persist_state_as_code -S "$1" foo; }
     cleanup(){
       local state_var="$1"
-      local foo="" bar=""
+      local foo bar
       hs_read_persisted_state -S "$state_var" foo bar
       printf "%s:%s" "$foo" "${bar:-}"
     }
@@ -611,7 +626,7 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
     init(){ local foo=secret; hs_persist_state_as_code -S "$1" foo; }
     cleanup(){
       local state_var="$1"
-      local foo="" bar="" baz=""
+      local foo bar baz
       hs_read_persisted_state -S "$state_var" foo bar baz
       printf "%s:%s:%s" "$foo" "${bar:-}" "${baz:-}"
     }
@@ -632,7 +647,7 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
     init(){ local foo=secret; hs_persist_state_as_code -S "$1" foo; }
     cleanup(){
       local state_var="$1"
-      local foo="" bar=""
+      local foo bar
       hs_read_persisted_state -q -S "$state_var" foo bar
       printf "%s:%s" "$foo" "${bar:-}"
     }
@@ -652,7 +667,7 @@ export -f make_state corrupt_state # makes it available in bash --noprofile -lc 
     init(){ local foo=secret; hs_persist_state_as_code -S "$1" foo; }
     cleanup(){
       local state_var="$1"
-      local foo="" bar=""
+      local foo bar
       hs_read_persisted_state -S "$state_var" -q foo bar
       printf "%s:%s" "$foo" "${bar:-}"
     }

